### DISCUSSION TITLE
Contribution

### DISCUSSION THREAD

**Discussion by Almesi** [sancarn/stdVBA#103] [General]
Hey there, like i said yesterday i would like to contribute. I started some time ago with my own Standard library. https://github.com/Almesi/AL_StdLib
If you find anything in there that is worth your interest i would gladly help you. I also got some other ideas which may be interesting.
Do you have a policy of what you would like to include in your Library and what not?

**Comment by sancarn** [comment]
Hi Almesi,

I've added [Contribution.md](https://github.com/sancarn/stdVBA/blob/master/contribution.md) which should help with some generic coding style guidance.

As for things I'm looking for:

* Data structures like `Stack`, `Heap`, `BinaryTree`, `QuadTree`, `R-tree`, `Graph`, `Digraph` will always be beneficial.
    * Additionally algorithms on top of these like Graph: A* search, Min spanning tree, shortest path, max flow etc.
* Linear Algebra stuff like `Matrix`, `Vector`, Cross product, Dot product, ... may be of benefit. Would maybe be worth looking at using OpenSL for some of these.

There is also a [roadmap](https://github.com/users/sancarn/projects/3/views/1) which you can look on to see if there are any specifics which interest you :)

**Comment by sihlfall** [comment]
Hi sancarn,
thank you for all your work on this great project. I see you are looking for a native Regex implementation. I do have one, which I could contribute. It's rather unpolished yet, but I could bring it to a good shape, if you like.
With best regards,
sihlfall

**Reply by sancarn** [reply]
@sihlfall That would be awesome :) You can look at [stdRegex2](https://github.com/sancarn/stdVBA/blob/master/src/WIP/stdRegex2.cls) for my general plans for regex implementation. So far parsing the patterns is done, and I have data generation half-done. Though it still needs some generic NFA approach really to work fully.

It also might be better to not use dictionaries in the long run, and opt for using UDTs instead... Not sure. Happy for anyone to transform stdRegex2 into whatever it needs to be :)

**Reply by sihlfall** [reply]
Dear sancarn,
now I have made a first version, here:

[https://github.com/sihlfall/stdVBA/tree/regex-vbaonly](https://github.com/sihlfall/stdVBA/tree/regex-vbaonly)

You can find the added files in

src/WIP/stdRegex3
tests/WIP/stdRegex3

I was not able to get down to a single class module (and I really do prefer not to); I need a class module as well as a standard module.

There isn't much documentation on the engine yet, as it's a single-file version of a regex engine I'm developing, which is not yet a public project. It all started with a VBA translation of the regex engine from the [duktype project](https://github.com/svaarala/duktape), but it has deviated significantly.

Several things are not yet supported (but not necessarily difficult to implement):
* case-insensitive matching
* multi-line matching
* global matching
* named captures

I'll keep working on this. Feedback is very welcome.
Best regards,
Sihlfall

**Reply by sancarn** [reply]
@sihlfall I'm not sure I fully see the issue with making this a single file? As far as I can tell there is nothing in the module which particularly warrants a separate `.bas` module? 

Looks good on the whole. Most of the flags should be relatively easy to implement as you say. This will be a massive improvement to the project! Thanks a tonne for doing this 🤯

Don't suppose you know if this uses NFA or DFA or neither?

**Reply by sihlfall** [reply]
@sancarn Thanks for your reply. It might be possible to make this a single cls. There are two issues that need to be considered: (1.) The engine needs Unicode tables, which need to be initialized. Obviously we do not want to have separate tables for each regex instance. So we cannot just copy and paste the whole bas content into the cls. (2.) I do not know how standard procedure calls differ from method calls in terms of performance. I'll have to measure.

The engine uses a depth-first search, so neither a NFA nor a DFA. A pure DFA approach cannot (? I'm not entirely sure) support captures, and NFAs (definitely) cannot support backreferences, see here: [https://swtch.com/~rsc/regexp/regexp1.html](https://swtch.com/~rsc/regexp/regexp1.html). It is unclear to me whether NFAs can support lookaheads or lookbehinds.

It might make sense to add a DFA- or NFA-based engine for regexs that do not contain backreferences, but I am not sure I'll find the motivation to do that.

**Reply by sancarn** [reply]
@sihlfall
At #1 - This is a very good point. Usually to get around this I create a protected method on the class, which creates a cached object and returns this on demand to instances. However a direct return would indeed still create copies of the array. I think the best alternative here is for the protected method to return a byref variant (`VT_BYREF`) to the superclass instance. @cristianbuse shows an [implementation of ref-array creation here](https://stackoverflow.com/a/71081452/6302131)

As such my suggestion would be along the lines of:

```vba
Type TSingleton
  UnicodeCanonLookupTable() As Integer
  UnicodeCanonRunsTable() As Long
End Type
Type TLookups
  UnicodeCanonLookupTable as Variant
  UnicodeCanonRunsTable as Variant
End Type
Type TThis
  Singleton as TSingleton
  Lookups as TLookups
End Type
Private This as TThis

Public Sub protInit()
  This.Lookups.UnicodeCanonLookupTable = stdRegex.protGetUnicodeCanonLookupTable()
  This.Lookups.UnicodeCanonRunsTable = stdRegex.protGetUnicodeCanonRunsTable()
End Sub
Public Function protGetUnicodeCanonLookupTable() as Variant
  if ubound(This.Singleton.UnicodeCanonLookupTable) = -1 then
    Redim This.Singleton.UnicodeCanonLookupTable(0 to 65535)
    '...populate...
  end if
  protGetUnicodeCanonLookupTable = getByRefArray(This.Singleton.UnicodeCanonLookupTable(0))
End Function
Public Function protGetUnicodeCanonRunsTable() as Variant
  if ubound(This.Singleton.UnicodeCanonRunsTable) = -1 then
    Redim This.Singleton.UnicodeCanonRunsTable(0 to 302)
    '...populate...
  end if
  protGetUnicodeCanonRunsTable= getByRefArray(This.Singleton.UnicodeCanonRunsTable(0))
End Function
```

At #2 - [See this](https://sancarn.github.io/vba-articles/performance-tips.html#s8-module-vs-class) Object and module member calls take the same amount of time to execute. Object creation takes more time than type creation, but this isn't really important imo given that regex is compile first then execute many times.

> The engine uses a depth-first search, so neither a NFA nor a DFA

Didn't know about a DFS approach. You are correct DFA cannot support capture groups. NFAs can support backreferences but not in a particularly efficient way. I think you may have misread the article.

> It might make sense to add a DFA- or NFA-based engine for regexs that do not contain backreferences, but I am not sure I'll find the motivation to do that.

Totally, imo it's not really necessary. A DFA approach for matching alone on a limited feature set would provide performance improvements but realistically this is a VBA project xD. We mainly just want to cover ourselves from the compatibility gap provided by VBScript, and if we can improve the feature-set too that would be amazing. Performance improvements may not even be significant enough to warrant it. I believe Rust's regex engine uses this [hybrid approach](https://github.com/rust-lang/regex/issues/66).

**Reply by sihlfall** [reply]
@sancarn Thanks for the reply and all the references, I'll check them out. Regarding NFAs and backreferences: Since a language defined by a regex with backreferences need not be regular, but a language defined by an NFA is necessarily regular, regexes with backreferences cannot generally be modelled by an NFA. See e.g. here: [https://www.npopov.com/2012/06/15/The-true-power-of-regular-expressions.html](https://www.npopov.com/2012/06/15/The-true-power-of-regular-expressions.html) Of course, this does not preclude some sort of hybrid approach.

**Reply by sihlfall** [reply]
@sancarn I have made some progress. In particular, my regex engine now supports named capturing groups. I have been able to create a single-file object module version similar to stdRegex. I am now putting the static data tables into the regex.bytecode field of the class instance (slightly abusing it) and call into the class instance for running the previously static functions. This allows me to keep the array typed.

The object module is here: [https://github.com/sihlfall/stdVBA/tree/regex-vbaonly/src/WIP/stdRegex3](https://github.com/sihlfall/stdVBA/tree/regex-vbaonly/src/WIP/stdRegex3)

And my regex engine (with some, yet not extensive documentation) is here: [https://github.com/sihlfall/vba-regex](https://github.com/sihlfall/vba-regex)

The whole thing should be usable. I did some extensive testing, yet mainly on the static version. I used some of the RE2 test cases for testing, yet they cover only a subset of features.

Feedback and bug reports are welcome.

**Reply by sancarn** [reply]
@sihlfall WOW! This is looking incredible :) I'll try to do some testing myself over the next few days but so far looks great! 👍

**Reply by sancarn** [reply]
@sihlfall This is some incredible stuff, really well done 👍 Taught me a few things too! 😅 Never knew you could pass UDTs to `Friend` methods. A minor suggestion though:

Instead of:

```vb
Private Const REGEX_ERR = vbObjectError + 3000
Private Const REGEX_ERR_INVALID_REGEXP_ESCAPE = REGEX_ERR + 1
Private Const REGEX_ERR_INVALID_ESCAPE = REGEX_ERR + 2
Private Const REGEX_ERR_INVALID_RANGE = REGEX_ERR + 3
Private Const REGEX_ERR_UNTERMINATED_CHARCLASS = REGEX_ERR + 4
Private Const REGEX_ERR_INVALID_REGEXP_GROUP = REGEX_ERR + 6
Private Const REGEX_ERR_INVALID_REGEXP_CHARACTER = REGEX_ERR + 7
Private Const REGEX_ERR_INVALID_QUANTIFIER = REGEX_ERR + 8
Private Const REGEX_ERR_UNEXPECTED_END_OF_PATTERN = REGEX_ERR + 9
Private Const REGEX_ERR_UNEXPECTED_REGEXP_TOKEN = REGEX_ERR + 10
Private Const REGEX_ERR_UNEXPECTED_CLOSING_PAREN = REGEX_ERR + 11
Private Const REGEX_ERR_INVALID_QUANTIFIER_NO_ATOM = REGEX_ERR + 12
Private Const REGEX_ERR_INVALID_IDENTIFIER = REGEX_ERR + 13
Private Const REGEX_ERR_INTERNAL_LOGIC_ERR = REGEX_ERR + 20
Private Const REGEX_ERR_INVALID_FLAG_ERR = REGEX_ERR + 30
Private Const REGEX_ERR_INVALID_REPLACEMENT_STRING = REGEX_ERR + 31
Private Const REGEX_ERR_INVALID_INPUT = REGEX_ERR + 50
```

Can you do the following instead?

```vb
Private Enum REGEX_ERR_TYPE
  REGEX_ERR = vbObjectError + 3000
  REGEX_ERR_INVALID_REGEXP_ESCAPE = REGEX_ERR + 1
  REGEX_ERR_INVALID_ESCAPE = REGEX_ERR + 2
  REGEX_ERR_INVALID_RANGE = REGEX_ERR + 3
  REGEX_ERR_UNTERMINATED_CHARCLASS = REGEX_ERR + 4
  REGEX_ERR_INVALID_REGEXP_GROUP = REGEX_ERR + 6
  REGEX_ERR_INVALID_REGEXP_CHARACTER = REGEX_ERR + 7
  REGEX_ERR_INVALID_QUANTIFIER = REGEX_ERR + 8
  REGEX_ERR_UNEXPECTED_END_OF_PATTERN = REGEX_ERR + 9
  REGEX_ERR_UNEXPECTED_REGEXP_TOKEN = REGEX_ERR + 10
  REGEX_ERR_UNEXPECTED_CLOSING_PAREN = REGEX_ERR + 11
  REGEX_ERR_INVALID_QUANTIFIER_NO_ATOM = REGEX_ERR + 12
  REGEX_ERR_INVALID_IDENTIFIER = REGEX_ERR + 13
  REGEX_ERR_INTERNAL_LOGIC_ERR = REGEX_ERR + 20
  REGEX_ERR_INVALID_FLAG_ERR = REGEX_ERR + 30
  REGEX_ERR_INVALID_REPLACEMENT_STRING = REGEX_ERR + 31
  REGEX_ERR_INVALID_INPUT = REGEX_ERR + 50
End Enum
```

And similarly with all your constants. Mainly because currently the locals window looks like this:

![image](https://github.com/user-attachments/assets/2a9b8443-2fb0-479d-b54e-0ddff9fef2a8)

**Reply by sihlfall** [reply]
I didn't know, either. But then somewhere very hidden on Stackoverflow, I detected a hint. And then I realized that the class is already making use of the fact when passing `This` to protInit.

I agree, that watch window is not convenient. Please see my new file, it should be better now.

**Reply by sancarn** [reply]
Awesome @sihlfall, just wondered, what do you think of generating data from regex? Would it be easy to add such methods to your engine?

Also, I love how this file is generated. Very cool :)

**Reply by sihlfall** [reply]
What do you mean by generating data? Can you give an example?

**Reply by sancarn** [reply]
Sure, so this is something I started doing in [stdRegex2](https://github.com/sancarn/stdVBA/blob/master/src/WIP/stdRegex2.cls#L73-L76). E.G.

```vb
With stdRegex.Create("[A-Z]-\d{5}")
  Debug.Print .Generate()  'F-46215
  Debug.Print .Generate()  'K-95642
  ...
End With
```

Mainly useful for creating dummy datasets. The code I was using for this is [moderately simple](https://github.com/sancarn/stdVBA/blob/master/src/WIP/stdRegex2.cls#L651-L784) but not 100% sure how easily this could be applied to your own engine.

**Reply by sihlfall** [reply]
I'd say (wild guess)
* it's simple for regexs without asserts (^, $, \b etc.) or lookarounds. You just traverse the abstract syntax tree left-to-right,
* it's a bit more tricky for regexs with asserts;
* it's possible, but a huge amount of work for regexs with lookarounds, yet without backreferences: You can transform the regex with lookarounds into a regex without lookarounds by intersecting NFAs. I remember that was discussed somewhere on Stackoverflow.
* I'd say it's nearly impossible for regexs with both lookarounds and backreferences, as in general, regexs with backreferences cannot be represented as an NFA, so we cannot get rid of the lookarounds. On the other hand, I've nowhere seen any theoretical result saying that lookarounds make regexs with backreferences more powerful, hence my statement is really only a wild guess.

The more practical problem I see is that you have to specify some probability distribution if there is more than one alternative. E.g., with what probabilities should `a`, `aa`, `aaa`, `aaaa` etc. be generated if the regex is `a+`? And should `.` really generate every possible Unicode character?

**Reply by sancarn** [reply]
@sihlfall Is there any way that I can help push this to a PR? I've been a little out of the loop. Looking at https://github.com/sihlfall/vba-regex/ everything looks ready? It just needs compiling and a PR submitting? Or is that here ready to go? https://github.com/sihlfall/stdVBA/blob/regex-vbaonly/src/WIP/stdRegex3/stdRegex3.cls

If so happy for you to promote this to `/src` and push a PR :) It'd be useful if we can make a github action to compile and automate any future PRs too :)

**Reply by sihlfall** [reply]
Dear sancarn,
sorry for not answering earlier, I was totally under water with other projects.
I try to keep stdRegex3.cls in sync with the `master` branch of vba-regex, so either one should be fine to use. The `dev` branch of vba-regex is a bit ahead (in particular, it supports unbounded modifiers). I am planning to do a merge shortly, after running all the tests.
Feel free to go with any one of these, I am pretty confident that the project is in a good state.
Best regards,
sihlfall
